core/block_validator.go

```diff
@@ -0,0 +1,248 @@
+// false positives where a header is present but the state is not.
+func (v *BlockValidator) ValidateBlock(block *types.Block) error {
+	if v.bc.HasBlock(block.Hash()) {
+		if _, err := state.New(block.Root(), v.bc.chainDb); err == nil {
+			return &KnownBlockError{block.Number(), block.Hash()}
+		}
+	}
+	parent := v.bc.GetBlock(block.ParentHash())
+	if parent == nil {
+		return ParentError(block.ParentHash())
+	}
+	if _, err := state.New(parent.Root(), v.bc.chainDb); err != nil {
+		return ParentError(block.ParentHash())
+	}
+
+	header := block.Header()
+	// validate the block header
+	if err := ValidateHeader(v.Pow, header, parent.Header(), false, false); err != nil {
+		return err
+	}
+	// verify the uncles are correctly rewarded
+	if err := v.VerifyUncles(block, parent); err != nil {
+		return err
+	}
+
+	// Verify UncleHash before running other uncle validations
+	unclesSha := types.CalcUncleHash(block.Uncles())
+	if unclesSha != header.UncleHash {
+		return fmt.Errorf("invalid uncles root hash. received=%x calculated=%x", header.UncleHash, unclesSha)
+	}
+
+	// The transactions Trie's root (R = (Tr [[i, RLP(T1)], [i, RLP(T2)], ... [n, RLP(Tn)]]))
+	// can be used by light clients to make sure they've received the correct Txs
+	txSha := types.DeriveSha(block.Transactions())
+	if txSha != header.TxHash {
+		return fmt.Errorf("invalid transaction root hash. received=%x calculated=%x", header.TxHash, txSha)
+	}
+
+	return nil
+}
+
+// ValidateState validates the various changes that happen after a state
+// transition, such as amount of used gas, the receipt roots and the state root
+// itself. ValidateState returns a database batch if the validation was a succes
+// otherwise nil and an error is returned.
+func (v *BlockValidator) ValidateState(block, parent *types.Block, statedb *state.StateDB, receipts types.Receipts, usedGas *big.Int) (batch ethdb.Batch, err error) {
+	header := block.Header()
+	if block.GasUsed().Cmp(usedGas) != 0 {
+		return nil, ValidationError(fmt.Sprintf("gas used error (%v / %v)", block.GasUsed(), usedGas))
+	}
+	// Validate the received block's bloom with the one derived from the generated receipts.
+	// For valid blocks this should always validate to true.
+	rbloom := types.CreateBloom(receipts)
+	if rbloom != header.Bloom {
+		return nil, fmt.Errorf("unable to replicate block's bloom=%x", rbloom)
+	}
+
+	// Tre receipt Trie's root (R = (Tr [[H1, R1], ... [Hn, R1]]))
+	receiptSha := types.DeriveSha(receipts)
+	if receiptSha != header.ReceiptHash {
+		return nil, fmt.Errorf("invalid receipt root hash. received=%x calculated=%x", header.ReceiptHash, receiptSha)
+	}
+
+	// Commit state objects/accounts to a database batch and calculate
+	// the state root. The database is not modified if the root
+	// doesn't match.
+	root, batch := statedb.CommitBatch()
+	if header.Root != root {
+		return nil, fmt.Errorf("invalid merkle root: header=%x computed=%x", header.Root, root)
+	}
+	return batch, nil
+}
```

Comment:
Moving it to BlockChain is what I am suggesting.It can still be in a function, but shouldn't be an interface method.